controller/internal/enforcer/acls/acl.go

```diff
@@ -1,100 +1,106 @@
+	cache ipprefix.IPcache
+}
+
 func newACL() *acl {
 	return &acl{
-		sortedPrefixLens: make([]int, 0),
-		prefixLenMap:     make(map[int]*prefixRules),
+		cache: ipprefix.NewIPCache(),
 	}
 }
 
-// acl holds all the ACLS in an internal DB
-type acl struct {
-	sortedPrefixLens []int
-	prefixLenMap     map[int]*prefixRules
-}
+var errNoMatch = errors.New("No Match")
 
-func (a *acl) reverseSort() {
+func (a *acl) addToCache(ip net.IP, mask int, port string, policy *policy.FlowPolicy) error {
+	var err error
+	var portList portActionList
 
-	// Get reverse sorted prefix lengths for reject rules
-	for k := range a.prefixLenMap {
-		a.sortedPrefixLens = append(a.sortedPrefixLens, k)
+	r, err := newPortAction(port, policy)
+	if err != nil {
+		return fmt.Errorf("unable to create port action: %s", err)
 	}
-	sort.Sort(sort.Reverse(sort.IntSlice(a.sortedPrefixLens)))
-}
-
-func (a *acl) addRule(rule policy.IPRule) (err error) {
-	ruleAdd := func(address, port string, policy *policy.FlowPolicy) error {
-		var subnet, mask uint32
 
-		parts := strings.Split(address, "/")
+	val, exists := a.cache.Get(ip, mask)
+	if !exists {
+		portList = portActionList{}
+	} else {
+		portList = val.(portActionList)
+	}
 
-		subnetSlice := net.ParseIP(parts[0])
-		if subnetSlice == nil {
-			return fmt.Errorf("invalid ip address: %s", parts[0])
-		}
+	portList = append(portList, r)
 
-		subnet = binary.BigEndian.Uint32(subnetSlice.To4())
+	a.cache.Put(ip, mask, portList)
 
-		maskValue := 0
+	return nil
+}
 
-		switch len(parts) {
-		case 1:
-			mask = 0xFFFFFFFF
-			maskValue = 32
+func (a *acl) matchRule(ip net.IP, port uint16, preReport *policy.FlowPolicy) (report *policy.FlowPolicy, packet *policy.FlowPolicy, err error) {
+	report = preReport
 
-		case 2:
-			maskValue, err = strconv.Atoi(parts[1])
-			if err != nil {
-				return fmt.Errorf("invalid address: %s", err)
-			}
+	err = errNoMatch
 
-			if mask > 32 {
-				return fmt.Errorf("invalid mask value: %d", mask)
-			}
-			mask = binary.BigEndian.Uint32(net.CIDRMask(maskValue, 32))
+	lookup := func(val interface{}) bool {
+		if val != nil {
+			portList := val.(portActionList)
 
-		default:
-			return fmt.Errorf("invalid address: %s", address)
-		}
-		plenRules, ok := a.prefixLenMap[maskValue]
-		if !ok {
-			plenRules = &prefixRules{
-				mask:  mask,
-				rules: make(map[uint32]portActionList),
+			report, packet, err = portList.lookup(port, report)
+			if err == nil {
+				return true
 			}
-			a.prefixLenMap[maskValue] = plenRules
 		}
+		return false
+	}
 
-		r, err := newPortAction(port, policy)
-		if err != nil {
-			return fmt.Errorf("unable to create port action: %s", err)
-		}
+	a.cache.RunIP(ip, lookup)
 
-		subnet = subnet & mask
-		plenRules.rules[subnet] = append(plenRules.rules[subnet], r)
-		return nil
-	}
+	return report, packet, err
+}
+
+func (a *acl) addRule(rule policy.IPRule) (err error) {
 
 	for _, proto := range rule.Protocols {
 		if strings.ToLower(proto) == constants.TCPProtoNum {
```

Comment:
this is a good candidate for - 
if x != y {
    continue
}

the following code should be a separate function.